### PR TITLE
Adding ExternallyInteractable

### PR DESCRIPTION
**Pull request by Sefiraat**
## Description

Something I am working on in other addons is the ability to trigger/interact with other SlimefunItems when placed in the world in a generic way, i.e. not relying on depending on each individual possible addon/specifying ID's to do so. 

## Proposed changes

The idea here is that addons can implement the Interface and, within the overridden method, complete whatever actions it would like to do when being interacted with and return a result object denoting a generic summary what happened as part of the interaction. 

Currently I have only included the success of the interaction and any possible items generated as part of said interaction (as this has been initially thought up for networks). If there are any ideas for other things to be added that would also be useful, I can whack that in as well.

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [ ] I have made sure that the proposed changes do not break compatibility across the supported Minecraft versions (1.16.* - 1.19.*).
- [ ] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:17]: I do not remember writing this code...
#identitytheft

Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:17]: Dr. CopyPasta strikes again xD

**Review by Sfiguz7 — DISMISSED** [review]
LGTM

**Comment by SchnTgaiSpock** [comment]
> If there are any ideas for other things to be added that would also be useful, I can whack that in as well.

Maybe if an interaction was unsuccessful, it could also return the reason? Like not researched, not enough of some resource, etc...; it could then be used for smth like a message to the player

**Review by Sfiguz7 — DISMISSED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:19]:
```suggestion
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:21]:
```suggestion
 */
@FunctionalInterface
public interface ExternallyInteractable {
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:37]: we should avoid inner classes as much as possible. Any reason we don't want to put this in its own class?

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:40]: Use case?

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:41]: I don't want this to turn into a huge class with 10 different possible outputs, may be worth making this more abstract so we don't need to refactor this later.
Not sure if I want to block on this yet.

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java
Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/ExternallyInteractable.java:40]: It just allows the object being interacted with to define a string message - could be anything. Then it's up to the interacting class how/if it deal withs it. 
So for example in my Light Controller block, it would just say something like 'Failed due to not having enough power' or 'Failed due to missing connection' and it could then be fed back to the user from the class that called for the interaction

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java:45]: Lets change this up (for some reason i cant do code suggestion but hey o

Lets change this to 
Returns wheter this result has a result message or is null.

This is code so null makes more sense then not

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java:47]: Do we want True to be capital T here since 'true' is never 'True'

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java:54]: add or null

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java:56]: again or null

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java:21]: or null again

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java:30]: or null again

**Review by J3fftw1 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java
Comment (J3fftw1) [inline src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java:29]: or null again

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java
Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/InteractionResult.java:56]: I won't put it here as that's the purpose of the annotation.

**Review by J3fftw1 — APPROVED** [review]

**Review by WalshyDev — APPROVED** [review]
(this was meant to be sent this morning, oops)

File: src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java
Comment (Sefiraat) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java:14]: It was in just the one class previously with a note from yourself to abstract it out, is there a middle ground, if so I may need something more specific in direction other than not liking it? Happy to make changes if needed, and likely before it's merged in to save potential breaking changes down the line, but I'm not sure what the mid-step here would be?

Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java:14]: I'm not completely sure what the actual solution is. Maybe a Consumer which is called and then returns whatever, generic to decide what object/whatever? I don't know.

I just don't like the fact we'd have to constantly add more and more and more.

Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/core/attributes/interactions/ItemInteractionResult.java:14]: I don't have the solution and that's why I'm fine with this for now. I can just feel us having to refactor later.

**Review by Sfiguz7 — APPROVED** [review]
Left some comments in Sefi's DMs, overall I share Walshy's opinion on all fronts.
As for the technical side, LGTM

**Pull request merged**
